### PR TITLE
Document the new coding style

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Update the coding standards document to the new coding style (including https://github.com/Mbed-TLS/mbedtls/pull/6836).

To be merged at the same time as the [new coding style](https://github.com/Mbed-TLS/mbedtls/issues/6359).

Fixes https://github.com/Mbed-TLS/mbedtls/issues/6603.

### CODE REVIEW

File: kb/development/mbedtls-coding-standards.md [hunk 1]
```diff
@@ -8,102 +8,101 @@ This document describes Mbed TLS preferences for code formatting, naming convent
...
+
 ```c
-    memset( buf, 0, sizeof( buf ) );
+    do {
+        x += f();
+        if (x == 0) {
+            continue;
+        }
+        x += g();
+    } while (condition);
```
Comment (mpg) [inline kb/development/mbedtls-coding-standards.md:133]: I think we should state explicitly somewhee that for function definitions, the opening brace is on its own line. This section looks like it's about braces placement in general, but so far its content only mentions one case: control statements.

(Btw, function definitions is no the only missing case: enum/struct/union, and array initializers also come to mind: they're mentioned in "Formatting of lists" below but I don't think anything's said explicitly about the placement of braces.)

File: kb/development/mbedtls-coding-standards.md [hunk 2]
```diff
@@ -8,6 +8,43 @@ This document describes Mbed TLS preferences for code formatting, naming convent
...
+```
+
+To enforce the coding style on specific files:
+```sh
+scripts/code_style.py --fix library/file_i_edited.c tests/suites/test_suite_i_also_edited.function
+```
+
+To enforce the coding style on all files checked into Git:
+```sh
+scripts/code_style.py --fix library/file_i_edited.c tests/suites/test_suite_i_also_edited.function
```
Comment (mpg) [inline kb/development/mbedtls-coding-standards.md:31]: Copy-pasta: should stop at `--fix`.

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good to me, just a couple minor points.

**Review by mpg — DISMISSED** [review]
LGTM! Thanks for writing the documentation in a way that's both clear, complete and concise!

File: kb/development/mbedtls-coding-standards.md [hunk 3]
```diff
@@ -8,102 +8,161 @@ This document describes Mbed TLS preferences for code formatting, naming convent
...
+* Inside the curly braces around initializers: `int a[] = { 1, 2, 3 };`
 
-Multiple related source code lines should be formatted to be easily readable:
+Do not put a space in the following places:
+
+* After the function or macro name in a function or macro call: `f(x, y)`
+* Inside parentheses: `if (f(sizeof(int), (int) y))`
+* Inside square brackets: `a[i + 1]`
+* Before a comma: `f(x, y)`
+* After a unary operator: `if (!condition)`, `++x`
```
Comment (athoelke) [inline kb/development/mbedtls-coding-standards.md:105]:
```suggestion
* After a prefix unary operator: `if (!condition)`, `++x`
```

File: kb/development/mbedtls-coding-standards.md [hunk 4]
```diff
@@ -8,102 +8,161 @@ This document describes Mbed TLS preferences for code formatting, naming convent
...
+* Before a unary operator: `x++`
```
Comment (athoelke) [inline kb/development/mbedtls-coding-standards.md:106]:
```suggestion
* Before a postfix unary operator: `x++`
```

File: kb/development/mbedtls-coding-standards.md [hunk 5]
```diff
@@ -210,18 +268,17 @@ Most functions should return `int`, more specifically `0` on success (the operat
...
 Function should avoid in-out parameters for length (multiplexing buffer size on entry with length used/written on exit) since they tend to impair readability. For example:
 ```c
-    mbedtls_write_thing( void *thing, unsigned char *buf, size_t *len ); // no
-    mbedtls_write_thing( void *thing, unsigned char *buf, size_t buflen,
-                         size_t *outlen ); // yes
+mbedtls_write_thing(..., unsigned char *buf, size_t *len); // no
+mbedtls_write_thing(..., unsigned char *buf, size_t buflen, size_t *outlen); // yes
 ```
 
 For PSA functions, [input buffers](https://armmbed.github.io/mbed-crypto/html/overview/conventions.html#input-buffer-sizes) have a `size_t xxx_size` parameter after the buffer pointer, and [output buffers](https://armmbed.github.io/mbed-crypto/html/overview/conventions.html#output-buffer-sizes) have a `size_t xxx_size` parameter for the buffer size followed by a `size_t *xxx_length` parameter for the output length. This convention is also preferred in new `mbedtls_xxx` code, but older modules often use different conventions.
```
Comment (athoelke) [inline kb/development/mbedtls-coding-standards.md:275]: Can we link to the canonical PSA API pages instead? But maybe that should be a different PR...

The 1.1 specification is at https://arm-software.github.io/psa-api/crypto/1.1/

Comment (gilles-peskine-arm) [reply kb/development/mbedtls-coding-standards.md:275]: any update on having a `latest` link?

**Review by mpg — APPROVED** [review]
Still looks good to me with the changes since my previous review.
